### ISSUE TITLE
[no-conditional-in-test] - Ban optional Chaining

### ISSUE THREAD

**Issue by christopher-buss** [jest-community/eslint-plugin-jest#1928]
Optional chaining is the same as a conditional, but currently is not linted by the `no-conditional-in-test` rule.
I'm requesting this is added!

I'm happy to submit a PR for this, would just like to know if it is within scope of the rule.

**Comment by G-Rath** [comment]
yeah I think go for it - part of me is reluctant since Jest assertions don't assert in the eyes of TS so optional chaining can be useful for things that cannot be undefined or null in practice, but I don't actually think that is super common and there are other ways to deal with it.

Worst case too we just revert or have an option or something like that 😄

**Comment by erunion** [comment]
It would be nice if this were opt-in instead because it's pretty common to do things like:

```ts
expect(object?.property).toBe('something');
```

Conditionally checking if `object` exists or not here doesn't matter here because if it doesn't the the assertion is always going to fail.

Same issue applies with the test cases in the submitted PR:

```ts
const values = something.map(thing => thing?.foo);

it('valid', () => {
  expect(values).toStrictEqual(['foo']);
});
```

If `values` is `[undefined]` then the test will fail.

What is the rule adding here that isn't already covered?

**Comment by G-Rath** [comment]
@erunion per my previous comment, I don't outright disagree so I'm keen to here more, but currently all the situations I'd consider common where you'd want to use optional chaining in a test I can think of alternatives that feel better (or at least acceptable)

The first example you provided can be replaced with `toHaveProperty`, and the second one feels correct so I'm not sure what you mean? If you're saying you'd like to have that `map` _in_ the test, I would agree but arguably you could do something like `toMatchInlineSnapshot`.

Also, there is disabling the rule inline which I expect is something you'll have to do from time to time with this kind of rule - there is a reason why it's not in our recommended config.

**Comment by christopher-buss** [comment]
Personally to me the fact that 'expect(object?.property).toBe('something');' is testing both the fact that object could be undefined + the property could be something would be better off defined as two tests since you _could_ be masking issues accidentally by not doing a defined check first. I can't think off the top of my head a reason why you'd want to test something that could potentially be undefined in a test. In this case anyway I'd myself always use a !

**Comment by lukpsaxo** [comment]
we ban the use of the `!` operator and use `?.` frequently in tests to avoid a typescript error (that would need a explicit cast or suppress error).

We could allow the use of `!` only in tests, but then it gets confusing, where am I allowed to use it.

I too would strongly favor this being opt-in, otherwise I think we have to not use this rule.

**Comment by G-Rath** [comment]
@lukpsaxo could you provide some examples of cases where you've not been able other Jest marchers in place of optional chains?

Fwiw, I am leaning towards just reverting this completely, but would still like to hear more examples first

**Comment by christopher-buss** [comment]
I have `!` banned as well (through eslint), but I have an exception for test files specifically.

@G-Rath If it's an issue for lots of people can we make it an option to enable instead?

**Comment by erunion** [comment]
My example was probably poorly contrived but we encounter this pretty often in areas of our OpenAPI schema tooling codebase where properties can often be a mix of `JSONSchema7`, `ReferenceObject` or `undefined`.

Absolute worst case scenario from our codebase that is already running with the TS `strict` compiler flag enabled:

```ts
const authUnauthorizedResponse = definition.components?.responses?.authUnauthorize as ResponseObject;

expect(
  (
    (
      (authUnauthorizedResponse?.content?.['application/json']?.schema as SchemaObject)
        ?.oneOf?.[0] as SchemaObject
    )?.allOf?.[0] as SchemaObject
  ).properties?.docs,
).toStrictEqual({
  type: 'string',
  format: 'url',
  description: expect.stringContaining('log URL where you can see more information'),
  example: 'https://docs.readme.com/logs/6883d0ee-cf79-447a-826f-a48f7d5bdf5f',
});
```

We can move these to non-null assertions but we also ban those everywhere.

Personally I've always looked at `no-conditional-in-test` as a way to sniff out cases where you are doing exclusionary _flow_ checks within tests, where those are easily prone to breakages. Optionally chaining object properties to me doesn't really fall under the same umbrella because the value you're asserting this data to hasn't changed and if `authUnauthorizedResponse?.content` is `undefined` or not doesn't really matter here just as long as it all equates to `{ type: string, ... }`.

I'm not wholly opposed to having a rule to restrict optional chains, I just don't think it should be the default behavior of this rule.

**Comment by G-Rath** [comment]
@erunion have you consided using `toHaveProperty` for that case? I'm pretty sure you can even strongly type that still with generics.

Still I agree with your view on the rule - let's revert or get this behind an option
